File: server/src/main/java/io/microconfig/server/vault/credentials/KubernetesTokenCredentials.java
package io.microconfig.server.vault.credentials;

import io.microconfig.server.vault.exceptions.VaultAuthException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import java.net.URI;
import java.net.http.HttpRequest;

import static io.microconfig.server.util.HttpUtil.httpSend;
import static io.microconfig.server.util.JsonUtil.objectNode;
import static io.microconfig.server.vault.VaultUtil.validateResponse;
import static java.net.http.HttpRequest.BodyPublishers.ofString;
import static java.net.http.HttpRequest.newBuilder;
import static java.time.Duration.ofSeconds;

@RequiredArgsConstructor
@Slf4j
public class KubernetesTokenCredentials implements VaultCredentials {
    private final String address;
    private final String path;
    private final String role;
    private final String jwt;
    private String token;

    @Override
    public String getToken() {
        if (token != null) return token;
        validate();
        var request = request();
        var response = httpSend(request);
        var node = validateResponse(response);
        var policy = node.path("auth").path("token_policies").toString();

        log.debug("Fetched token with k8s JWT. Assigned policies: {}", policy);
        token = node.path("auth").path("client_token").asText();
        return token;
    }

    private void validate() {
        if (address == null) throw new VaultAuthException("Missing microconfig.vault.address");
        if (path == null) throw new VaultAuthException("Missing microconfig.vault.kubernetes.path");
        if (role == null) throw new VaultAuthException("Missing microconfig.vault.kubernetes.role");
        if (jwt == null) throw new VaultAuthException("Missing microconfig.vault.kubernetes.jwt");
    }

    private HttpRequest request() {
        var body = objectNode()
                .put("role", role)
                .put("jwt", jwt)
                .toString();

        var path = String.format("%s/v1/auth/%s/login", address, this.path);
        log.debug("Calling {} to auth with kubernetes jwt for '{}' role", path, role);
        return newBuilder(URI.create(path))
                .POST(ofString(body))
                .timeout(ofSeconds(10))
                .build();
    }
}

File: server/src/main/kotlin/io/microconfig/server/vault/credentials/KubernetesTokenCredentials.kt
package io.microconfig.server.vault.credentials

import io.microconfig.server.common.POST
import io.microconfig.server.common.httpClient
import io.microconfig.server.common.logger
import io.microconfig.server.common.send
import io.microconfig.server.common.toJson
import io.microconfig.server.vault.vaultResponse
import java.net.http.HttpRequest
import java.time.Duration

class KubernetesTokenCredentials(
    val address: String,
    val timeout: Duration,
    val path: String,
    val role: String,
    val jwt: String
) : VaultCredentials {

    private val log = logger()
    private val http = httpClient()
    private var token: String? = null

    override fun getToken(): String {
        if (token != null) return token!!

        val node = request().send(http).vaultResponse()
        token = node.path("auth").path("client_token").asText()
        log.debug("Fetched token with k8s role $role")
        return token!!
    }

    private fun request(): HttpRequest {
        val body = Request(role, jwt).toJson()
        val url = "$address/v1/auth/$path/login"
        return POST(url, body)
            .timeout(timeout)
            .build()
    }

    data class Request(
        val role: String,
        val jwt: String
    )

}